### PR TITLE
Improve topic messages by topic id query performance

### LINKED ISSUE #2694: /topics/{id}/messages response time is high
### Description

`/topics/{id}/messages` is very slow and tends to time out for certain topics which have a lot of messages when the `topic_message` table reaches several tens of millions rows.

pg explain shows for such topics, pg planner always chooses the primary key (just on `consensus_timestamp`) for an index scan, rather than the much more performant index on `(realm_num, topic_num, consensus_timestamp)`.

pg planner makes the decision based on the cost estimation and for such topics, the cost estimation of the better index is more than ten times of that when using the primary key.

There are 3 options to force pg planner to use the correct index:

1. set random_page_cost = 0, just for such topic messages queries. why it works is complicated, in short, it tricks the pg cost estimation and makes the cost on the better index lower than that of the primary key.

2. enable more frequent autovacuum on table `topic_message`, and rewrite the query:

   ```sql
   select *
   from topic_message
   where consensus_timestamp in (
     select consensus_timestamp
     from topic_message
     where realm_num = 0 and topic_num = <topic_num>
     order by consensus_timestamp
     limit 500
   )
   order by consensus_timestamp
   ```
   
   The query works in the way that if the table's visibility map is up-to-date (done by vacuum), the sub query will be an index-only scan on `(realm_num, topic_num, consensus_timestamp)`. However, there'll be times pg planner choose not to use index-only scan    and again falls back to the primary key, because the visibility map isn't up-to-date.

3. Use https://pghintplan.osdn.jp/pg_hint_plan.html, and add a hint section to the query.

### Steps to reproduce

N/A

### Additional context

_No response_

### Hedera network

testnet

### Version

v0.42.0

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by xin-hedera**
**Description**:

This PR fixes the topic messages by id query performance issue.

- set random_page_cost = 0 for the topic messages by id query

**Related issue(s)**:

Fixes #2694

**Notes for reviewer**:

Without the optimization, querying topic messages for topic `0.0.120438` in mainnet times out for both the rest api and grpc subscription.

With the optimization:

```
time curl -s -o /dev/null localhost:11000/api/v1/topics/120438/messages

real  0m0.073s
user  0m0.005s
sys 0m0.005s

2021-10-19T12:54:02.641-0600 INFO grpc-executor-9 c.h.m.g.s.TopicMessageServiceImpl [6SIEfnEV] Topic onComplete 0.0.120438 with 100 messages in 90.71 ms (1111/s)
```

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-rest/topicmessage.js [hunk 1]
```diff
@@ -270,14 +275,36 @@ const getMessage = async (pgSqlQuery, pgSqlParams) => {
...
+    } catch (err) {
+      if (client !== undefined) {
+        await client.query('rollback');
+      }
+      throw new DbError(err.message);
+    } finally {
+      if (client !== undefined) {
+        client.release();
+      }
+    }
```
Comment (xin-hedera) [inline hedera-mirror-rest/topicmessage.js:303]: this is how to do explicit transaction with `node-postgres`. We need to keep the client so the queries and the transaction all happen in the same connection / session.

File: hedera-mirror-rest/topicmessage.js [hunk 2]
```diff
@@ -25,9 +25,14 @@ const config = require('./config');
 const constants = require('./constants');
 const EntityId = require('./entityId');
 const utils = require('./utils');
+const {DbError} = require('./errors/dbError');
 const {NotFoundError} = require('./errors/notFoundError');
 const {InvalidArgumentError} = require('./errors/invalidArgumentError');
 
+// make the cost estimation of using the index on (topic_id, consensus_timestamp) lower than that of the primary key so
+// pg planner will choose the better index when querying topic messages by id
+const topicMessagesByIdQueryHint = 'set local random_page_cost = 0';
```
Comment (xin-hedera) [inline hedera-mirror-rest/topicmessage.js:34]: pg `set local` will fail if it's not run in a transaction

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java [hunk 3]
```diff
@@ -69,6 +73,7 @@
             typedQuery.setMaxResults((int) filter.getLimit());
         }
 
+        entityManager.createNativeQuery(TOPIC_MESSAGES_BY_ID_QUERY_HINT).executeUpdate();
```
Comment (xin-hedera) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java:76]: the method `findByFilter` is annotated with spring's `Transactional` in the interface, so the `set local` statement is run inside a transaction and won't fail.

**Review by Nana-EC — APPROVED** [review]
LGTM

**Pull request merged**